mmv1/third_party/terraform/services/bigquery/resource_bigquery_table.go.tmpl

```diff
@@ -2737,21 +2775,87 @@ func expandView(configured interface{}) *bigquery.ViewDefinition {
 	raw := configured.([]interface{})[0].(map[string]interface{})
 	vd := &bigquery.ViewDefinition{Query: raw["query"].(string)}
 
-	if v, ok := raw["use_legacy_sql"]; ok {
-		vd.UseLegacySql = v.(bool)
+	{{- if ne $.TargetVersionName "ga" }}
+	if v, ok := raw["foreign_definitions"]; ok && len(v.([]interface{})) > 0 {
```

Comment:
Having a field that goes from defaulting to true to then defaulting to true **ONLY** if a specific field is not set introduces some unexpected behavior to users. 

Although I understand your reasoning for this, it still wouldn't be possible to merge due to reasoning mentioned above. We'll want to merge this in the next Major Release. 

Any input on your end would be appreciated @c2thorn

Reply:
What does the API default to? Can we just use the API's default instead of a provider default by setting `use_legacy_sql` to optional + computed. If there is no API default, why do we have a provider default in the first place?

Alternatively, can we have users explicitly set `use_legacy_sql` to false in order to use the new field? Or will the API _only_ accept no value?

We definitely should not have a value change in an expander where it will not show during plan-time _especially_ for a non-Computed value. If we can't do the suggestions above, the best way to set a provider-level conditional default is to make the field Optional+Computed and move the condition logic to a CustomizeDiff

Reply:
It's relatively new and will need testing, but `d.GetRawConfig().GetAttr("field")` should get the actual config value. you may be able to see if it is equivalent to `cty.UnknownVal(cty.Bool)` to determine whether or not `use_legacy_sql` was set in config or unknown (left to an arbitrary computed value)

Reply:
Searching `GetRawConfig()` should show some uses of it elsewhere in Magic Modules. You'll probably need to pass in the `d` resourcedata object to the expander or move the logic elsewhere

Reply:
I updated `view.use_legacy_sql` to be Optional + Computed and cleaned up the related logic in the latest commit. Removing the default value is unavoidable though because the API only accepts no value for `use_legacy_sql` when `foreign_definitions` is set. Because Optional + Computed, existing configs shouldn't be impacted. Is there any way for this to be allowed in a minor release?

Reply:
@BBBmau @c2thorn This PR is ready for review again.

Reply:
Sorry @wj-chen, I was OOO until yesterday and still catching up. 

So `use_legacy_sql` went from default true to O+C with custom logic to set/unset based on the TF config value.

With the code from https://github.com/GoogleCloudPlatform/magic-modules/pull/12390/files#diff-d9d1047fa09e43341e0a3dfc5af24525b5859851dd6eaf51888d50934952dd8aR2783, what happens to existing configs that use the previous default of true when they upgrade?

If I'm following the new logic from correctly, we will no longer send any value for `vd.UseLegacySql` where before the upgrade we were sending `true`. This would still be a breaking change if so.

Are you able to test this scenario to confirm?

Reply:
> Because Optional + Computed, existing configs shouldn't be impacted. Is there any way for this to be allowed in a minor release?

If I'm following the logic correctly, existing configs won't be immediately affected because O+C, but the next time the expander runs (like in an update), the config value will be null and we will never send a value to the API which would unset the field on the backend?

Reply:
Just tested and there should be no impact to the existing users:

For users who has a config without `use_legacy_sql`, the current provider version sends `true` in the create request, and table gets created with `use_legacy_sql = true`.

Then after upgrading to using this code, `terraform apply` on the same config without change (i.e. still no `use_legacy_sql` specified) shows no diff. `terraform apply` on the config with change in another field (something not `use_legacy_sql`) shows a diff for only that change, and when run, that update is issued without `use_legacy_sql` in the request, and the API response leaves `use_legacy_sql = true` unchanged for the table.

The biggest impact with this change is that for new configs run with the new provider version, not specifying `use_legacy_sql` means not setting it to `true` unlike the current behavior. For users that want it to be true, they have to set it explicitly now. If this is ok, I will update the release note to call it out.

Reply:
There's a chance that some users expect the default to be true for new resources and may be surprised that they would now need to set it. From their perspective, it would be a breaking change that they now need to update their configs to get the same resource creation behavior as the previous minor version.

Thanks for confirming that existing resources wouldn't be affected. That takes this from "definitely can't do this" to "probably can't do this" from my perspective. I still think we could get around this last part by implementing a CustomizeDiff to conditionally set the default of true if `foreign_definitions` is not set

Reply:
Looks like we've done something similar here: https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/templates/terraform/constants/tpu_vm.go.tmpl#L33